homeassistant/components/hive/water_heater.py

```diff
@@ -88,18 +78,21 @@ def operation_list(self):
         """List of available operation modes."""
         return SUPPORT_WATER_HEATER
 
-    async def async_added_to_hass(self):
-        """When entity is added to Home Assistant."""
-        await super().async_added_to_hass()
-        self.session.entities.append(self)
-
+    @refresh_system
     def set_operation_mode(self, operation_mode):
         """Set operation mode."""
         new_mode = HASS_TO_HIVE_STATE[operation_mode]
         self.session.hotwater.set_mode(self.node_id, new_mode)
 
-        for entity in self.session.entities:
-            entity.handle_update(self.data_updatesource)
+    async def async_added_to_hass(self):
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
+        self.session.entity_lookup[self.entity_id] = self.node_id
+        dispatcher_connect(self.hass, DOMAIN, self._update_callback)
+
+    @callback
+    def _update_callback(self) -> None:
+        self.schedule_update_ha_state()
```

Comment:
Use `async_schedule_update_ha_state`. We're in async context since we're decorated with `@callback`.